packages/core/src/useKeyedChildListeners.tsx

```diff
@@ -1,31 +1,39 @@
 import * as React from 'react';
 
-import type { KeyedListenerMap } from './NavigationBuilderContext';
+import type {
+  KeyedListenerMap,
+  MapValueType,
+} from './NavigationBuilderContext';
+
+type KeyedChildListeners = {
+  [ListenerType in keyof KeyedListenerMap]: Map<
+    string | undefined,
+    MapValueType<KeyedListenerMap[ListenerType]> | undefined
+  >;
+};
 
 /**
  * Hook which lets child navigators add getters to be called for obtaining rehydrated state.
  */
 export default function useKeyedChildListeners() {
-  const { current: keyedListeners } = React.useRef<{
-    [K in keyof KeyedListenerMap]: Record<
-      string,
-      KeyedListenerMap[K] | undefined
-    >;
-  }>({
-    getState: {},
-    beforeRemove: {},
+  const { current: keyedListeners } = React.useRef<KeyedChildListeners>({
+    getState: new Map(),
+    beforeRemove: new Map(),
+    action: new Map(),
+    focus: new Map(),
   });
 
   const addKeyedListener = React.useCallback(
-    <T extends keyof KeyedListenerMap>(
-      type: T,
-      key: string,
-      listener: KeyedListenerMap[T]
+    <ListenerType extends keyof KeyedChildListeners>(
+      type: ListenerType,
+      key: string | undefined,
+      listener: MapValueType<KeyedListenerMap[ListenerType]> | undefined
     ) => {
-      keyedListeners[type][key] = listener;
+      const keyToSet = key !== undefined ? key : 'undefined';
```

Comment:
This is one area we're not sure about the best way to handle. 

https://github.com/react-navigation/react-navigation/pull/10596/files#diff-8186269a9d18bc03947df26c6d0acaabe753c63a7bf0bec13280b7f06aecd266R621 states that sometimes the route.key can be missing when calling useOnAction which means this could now be called without a key (whereas before it was going to useChildListeners which didn't need key).

For now, we've been assigning the hardcoded key of `undefined` which is probably not ideal. For someone who knows the codebase better; in what situations is route.key missing at the above-referenced line? Would there only ever be one listener without a key or would we need to cope with there potentially being multiple listeners with no key?